File: samples/dotnet/BgeM3.Onnx/M3EmbedderFactory.cs
using Microsoft.ML.OnnxRuntime;

namespace BgeM3.Onnx;

/// <summary>
/// Factory class for creating M3Embedder instances with common configurations
/// </summary>
public static class M3EmbedderFactory
{
    /// <summary>
    /// Creates an M3Embedder optimized for CPU inference
    /// </summary>
    /// <param name="tokenizerPath">Path to the ONNX tokenizer model</param>
    /// <param name="modelPath">Path to the ONNX embedding model</param>
    /// <returns>M3Embedder configured for CPU</returns>
    public static M3Embedder CreateCpuOptimized(string tokenizerPath, string modelPath)
    {
        var config = new M3EmbedderConfig
        {
            ExecutionProvider = ExecutionProvider.CPU,
            EnableMemoryPattern = true,
            EnableCpuMemArena = true,
            LogSeverityLevel = OrtLoggingLevel.ORT_LOGGING_LEVEL_WARNING
        };

        return new M3Embedder(tokenizerPath, modelPath, config);
    }

    /// <summary>
    /// Creates an M3Embedder optimized for CUDA inference
    /// </summary>
    /// <param name="tokenizerPath">Path to the ONNX tokenizer model</param>
    /// <param name="modelPath">Path to the ONNX embedding model</param>
    /// <param name="deviceId">CUDA device ID (default: 0)</param>
    /// <returns>M3Embedder configured for CUDA</returns>
    public static M3Embedder CreateCudaOptimized(string tokenizerPath, string modelPath, int deviceId = 0)
    {
        var config = new M3EmbedderConfig
        {
            ExecutionProvider = ExecutionProvider.CUDA,
            FallbackProviders = [ExecutionProvider.CPU],
            CudaDeviceId = deviceId,
            EnableMemoryPattern = true,
            EnableCpuMemArena = false, // Disable CPU memory arena when using GPU
            LogSeverityLevel = OrtLoggingLevel.ORT_LOGGING_LEVEL_WARNING
        };

        return new M3Embedder(tokenizerPath, modelPath, config);
    }

    /// <summary>
    /// Creates an M3Embedder with custom configuration
    /// </summary>
    /// <param name="tokenizerPath">Path to the ONNX tokenizer model</param>
    /// <param name="modelPath">Path to the ONNX embedding model</param>
    /// <param name="primaryProvider">Primary execution provider</param>
    /// <param name="fallbackProviders">Fallback providers in order of preference</param>
    /// <param name="cudaDeviceId">CUDA device ID (used only if CUDA is specified)</param>
    /// <returns>M3Embedder with custom configuration</returns>
    public static M3Embedder CreateCustom(
        string tokenizerPath,
        string modelPath,
        ExecutionProvider primaryProvider,
        ExecutionProvider[]? fallbackProviders = null,
        int cudaDeviceId = 0)
    {
        var config = new M3EmbedderConfig
        {
            ExecutionProvider = primaryProvider,
            FallbackProviders = fallbackProviders ?? [ExecutionProvider.CPU],
            CudaDeviceId = cudaDeviceId,
            EnableMemoryPattern = true,
            EnableCpuMemArena = primaryProvider == ExecutionProvider.CPU,
            LogSeverityLevel = OrtLoggingLevel.ORT_LOGGING_LEVEL_WARNING
        };

        return new M3Embedder(tokenizerPath, modelPath, config);
    }
}

File: samples/java/bge-m3-onnx/src/main/java/com/yunikosoftware/bgem3onnx/M3EmbedderFactory.java
package com.yunikosoftware.bgem3onnx;

import ai.onnxruntime.OrtException;

/**
 * Factory class for creating M3Embedder instances with common configurations
 */
public class M3EmbedderFactory {
    
    /**
     * Creates an M3Embedder optimized for CPU inference
     * 
     * @param tokenizerPath Path to the ONNX tokenizer model
     * @param modelPath Path to the ONNX embedding model
     * @return M3Embedder configured for CPU
     * @throws OrtException If there's an error initializing the ONNX sessions
     */
    public static M3Embedder createCpuOptimized(String tokenizerPath, String modelPath) throws OrtException {
        M3EmbedderConfig config = new M3EmbedderConfig.Builder()
                .executionProvider(ExecutionProvider.CPU)
                .enableMemoryPattern(true)
                .enableCpuMemArena(true)
                .logSeverityLevel(2)
                .build();

        return new M3Embedder(tokenizerPath, modelPath, config);
    }

    /**
     * Creates an M3Embedder optimized for CUDA inference
     * 
     * @param tokenizerPath Path to the ONNX tokenizer model
     * @param modelPath Path to the ONNX embedding model
     * @param deviceId CUDA device ID (default: 0)
     * @return M3Embedder configured for CUDA
     * @throws OrtException If there's an error initializing the ONNX sessions
     */
    public static M3Embedder createCudaOptimized(String tokenizerPath, String modelPath, int deviceId) throws OrtException {
        M3EmbedderConfig config = new M3EmbedderConfig.Builder()
                .executionProvider(ExecutionProvider.CUDA)
                .fallbackProviders(ExecutionProvider.CPU)
                .cudaDeviceId(deviceId)
                .enableMemoryPattern(true)
                .enableCpuMemArena(false) // Disable CPU memory arena when using GPU
                .logSeverityLevel(2)
                .build();

        return new M3Embedder(tokenizerPath, modelPath, config);
    }

    /**
     * Creates an M3Embedder optimized for CUDA inference with default device ID (0)
     * 
     * @param tokenizerPath Path to the ONNX tokenizer model
     * @param modelPath Path to the ONNX embedding model
     * @return M3Embedder configured for CUDA
     * @throws OrtException If there's an error initializing the ONNX sessions
     */
    public static M3Embedder createCudaOptimized(String tokenizerPath, String modelPath) throws OrtException {
        return createCudaOptimized(tokenizerPath, modelPath, 0);
    }

    /**
     * Creates an M3Embedder with custom configuration
     * 
     * @param tokenizerPath Path to the ONNX tokenizer model
     * @param modelPath Path to the ONNX embedding model
     * @param primaryProvider Primary execution provider
     * @param fallbackProviders Fallback providers in order of preference
     * @param cudaDeviceId CUDA device ID (used only if CUDA is specified)
     * @return M3Embedder with custom configuration
     * @throws OrtException If there's an error initializing the ONNX sessions
     */
    public static M3Embedder createCustom(String tokenizerPath, String modelPath, 
                                         ExecutionProvider primaryProvider, 
                                         ExecutionProvider[] fallbackProviders, 
                                         int cudaDeviceId) throws OrtException {
        M3EmbedderConfig config = new M3EmbedderConfig.Builder()
                .executionProvider(primaryProvider)
                .fallbackProviders(fallbackProviders != null ? fallbackProviders : new ExecutionProvider[]{ExecutionProvider.CPU})
                .cudaDeviceId(cudaDeviceId)
                .enableMemoryPattern(true)
                .enableCpuMemArena(primaryProvider == ExecutionProvider.CPU)
                .logSeverityLevel(2)
                .build();

        return new M3Embedder(tokenizerPath, modelPath, config);
    }
}